# I. Importing. Creating DF

1. (2б) Імпортувати бібліотеку pandas та numpy.

In [109]:
import pandas as pd
import numpy as np

2. (3б) Створити тестовий датафрейм зі списку, що містить записи пацієнтів про їхні прізвища, імена, дату народження, масу тіла та наявність медичного страхування. Одне зі значень зробить пустим. Ознайомтесь з типом кожної колонки.

In [110]:
task_2_df = pd.DataFrame([["Dragovoz", "Damian", "28-12-2004", 64, False],
                          ["Pines", "Dipper", "31-8-1999", 50, np.nan]],
                          columns=["Surname", "Name", "Birth date", "Weight", "Health care"])
print(task_2_df)

    Surname    Name  Birth date  Weight Health care
0  Dragovoz  Damian  28-12-2004      64       False
1     Pines  Dipper   31-8-1999      50         NaN


In [111]:
print(task_2_df.dtypes)

Surname        object
Name           object
Birth date     object
Weight          int64
Health care    object
dtype: object


# II. Reading from file

3. (7б) Завантажити файл missile_attacks_daily.csv з [Kaggle](https://www.kaggle.com/datasets/piterfm/massive-missile-attacks-on-ukraine?select=missile_attacks_daily.csv). Ознайомитись з описом датасету. Додати файл до папки поточної сесії colab notebook на панелі ліворуч або на гугл диск (локально до проєкту, якщо працюєте у пайчармі). Зчитати датафрейм з файлу та вивести датафрейм на екран. Зверніть увагу на те, щоб всі колонки відповідали наведеним у описі типам даних.

In [112]:
data_path = "D:/Admin/Desktop/Collage year 2/Python/pythonProject/missile_attacks_daily.csv"
df = pd.read_csv(data_path, sep=',', header=0, na_values='?')


# III. Data understanding

4. (2б) Виведіть на екран перші 10 рядків зчитаного датафрейму.

In [113]:
df.head(10)

time_start          time_end           model  \
0        2024-03-24  2024-03-24 08:00     X-101/X-555   
1        2024-03-24  2024-03-24 08:00  Shahed-136/131   
2        2024-03-23  2024-03-23 08:00  Shahed-136/131   
3        2024-03-23  2024-03-23 08:00           C-300   
4        2024-03-23        2024-03-23            ZALA   
5  2024-03-22 02:00  2024-03-22 09:00            X-59   
6  2024-03-22 02:00  2024-03-22 09:00    X-47 Kinzhal   
7  2024-03-22 02:00  2024-03-22 09:00            X-22   
8  2024-03-22 02:00  2024-03-22 09:00     X-101/X-555   
9  2024-03-22 02:00  2024-03-22 09:00  Shahed-136/131   

                            launch_place          target  \
0                               Engels-2         Ukraine   
1  Primorsko-Akhtarsk and Chauda, Crimea         Ukraine   
2        Chauda, Crimea and Kursk oblast         Ukraine   
3                                    NaN  Donetsk oblast   
4                                    NaN           south   
5                    Zaporizhzhia oblast            east   
6                          Tambov oblast         Ukraine   
7                          Rostov oblast         Ukraine   
8                            Caspian Sea         Ukraine   
9                     Primorsko-Akhtarsk         Ukraine   

                                   destroyed_details       carrier  launched  \
0                                                NaN  14 x Tu-95MS      29.0   
1  {'Dnipropetrovsk oblast': 17, 'Kherson oblast'...           NaN      28.0   
2  {'Poltava oblast': 2, 'Kherson oblast': 3, 'Dn...           NaN      34.0   
3                                                NaN           NaN       4.0   
4                                                NaN           NaN       2.0   
5                                                NaN         Su-34       2.0   
6                                                NaN  10 x MIG-31K       7.0   
7                                                NaN       Tu-22M3       5.0   
8                                                NaN  13 x Tu-95MS      40.0   
9  {'Odesa oblast': 14, 'Kherson oblast': 5, 'Dni...           NaN      63.0   

   destroyed                                             source  
0         18  kpszsu/posts/pfbid02E9itApFzseeXaKMzes3zT2DHau...  
1         25  kpszsu/posts/pfbid02E9itApFzseeXaKMzes3zT2DHau...  
2         31  kpszsu/posts/pfbid0Jo2zmM4BzAZfS91x3Eb5Aa6L7AF...  
3          0  kpszsu/posts/pfbid0Jo2zmM4BzAZfS91x3Eb5Aa6L7AF...  
4          2  PvKPivden/posts/pfbid02rMwctKZrMFpQUXBtK15WCD8...  
5          2  kpszsu/posts/pfbid0G2H3qnSqb4ZcDPgrD1by43GSeZ4...  
6          0  kpszsu/posts/pfbid0G2H3qnSqb4ZcDPgrD1by43GSeZ4...  
7          0  kpszsu/posts/pfbid0G2H3qnSqb4ZcDPgrD1by43GSeZ4...  
8         35  kpszsu/posts/pfbid0G2H3qnSqb4ZcDPgrD1by43GSeZ4...  
9         55  kpszsu/posts/pfbid0G2H3qnSqb4ZcDPgrD1by43GSeZ4...

5. (2б) Отримайте інформацію про форму датафрейму.

In [114]:
df.shape

(861, 10)

6. (3б) Отримайте дані про кожну колонку з її назвою, типом даних та кількістю непустих значень.

In [115]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 861 entries, 0 to 860
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   time_start         861 non-null    object 
 1   time_end           861 non-null    object 
 2   model              861 non-null    object 
 3   launch_place       466 non-null    object 
 4   target             833 non-null    object 
 5   destroyed_details  68 non-null     object 
 6   carrier            169 non-null    object 
 7   launched           858 non-null    float64
 8   destroyed          861 non-null    int64  
 9   source             861 non-null    object 
dtypes: float64(1), int64(1), object(8)
memory usage: 67.4+ KB


7. (3б) Отримайте статистику для колонок числового типу.

In [116]:
df.describe()

launched   destroyed
count  858.000000  861.000000
mean     9.083916    7.038328
std     12.747066   11.021002
min      1.000000    0.000000
25%      1.000000    1.000000
50%      4.000000    2.000000
75%     12.000000   10.000000
max     96.000000   87.000000

8. (2б) Отримати кількість унікальних значень у кожній колонці.

In [117]:
df.nunique()

time_start           436
time_end             482
model                 43
launch_place          85
target                79
destroyed_details     68
carrier               50
launched              57
destroyed             54
source               443
dtype: int64

9. (3б) Вивести на екран значення та кількість цих значень у колонці, яка має найменше унікальних значень (див. попереднє завдання).

In [118]:
print(df["model"].value_counts)

<bound method IndexOpsMixin.value_counts of 0             X-101/X-555
1          Shahed-136/131
2          Shahed-136/131
3                   C-300
4                    ZALA
              ...        
856              Orlan-10
857              Orlan-10
858        Shahed-136/131
859                  X-59
860    Reconnaissance UAV
Name: model, Length: 861, dtype: object>


10. (15б) На основі отриманої інформації про датасет вище, опишіть по пунктах, що б ви виправили або на що б звернули увагу на кроці очищення даних і чому?

 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   time_start         861 non-null    object 
 1   time_end           861 non-null    object 
 2   model              861 non-null    object 
 3   launch_place       466 non-null    object 
 4   target             833 non-null    object 
 5   destroyed_details  68 non-null     object 
 6   carrier            169 non-null    object 
 7   launched           858 non-null    float64
 8   destroyed          861 non-null    int64  
 9   source             861 non-null    object 

 		launched	destroyed
count	858.000000	861.000000
mean	9.083916	7.038328
std	12.747066	11.021002
min	1.000000	0.000000
25%	1.000000	1.000000
50%	4.000000	2.000000
75%	12.000000	10.000000
max	96.000000	87.000000

 1. launches не повинен бути float
 2. destriyed_details майже завжди не запоанено

...


# IV. Data Cleaning

11. (3б) Перетворити колонку launched у цілочислену (Int64). Перевірити, чи все пройшло, як очікувалось.

In [119]:
print(df.dtypes)


time_start            object
time_end              object
model                 object
launch_place          object
target                object
destroyed_details     object
carrier               object
launched             float64
destroyed              int64
source                object
dtype: object


In [120]:
df.astype({"launched": "Int64"})
print(df.dtypes)

time_start            object
time_end              object
model                 object
launch_place          object
target                object
destroyed_details     object
carrier               object
launched             float64
destroyed              int64
source                object
dtype: object


12. (3б) Видалити колонки, де кількість непустих значень становить менше 30% від загальної кількості записів. Можна вручну порахувати і тут тільки видалити. За автоматизацію розрахунку відстоків - додаткові 3 бали.

In [121]:
df.info()
print()
# 861 line total <- len(df)
# 30% = 258    

print('\n', df.isnull().mean(), '\n')
df = df.loc[:, df.isnull().mean() < 1 - 0.3]
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 861 entries, 0 to 860
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   time_start         861 non-null    object 
 1   time_end           861 non-null    object 
 2   model              861 non-null    object 
 3   launch_place       466 non-null    object 
 4   target             833 non-null    object 
 5   destroyed_details  68 non-null     object 
 6   carrier            169 non-null    object 
 7   launched           858 non-null    float64
 8   destroyed          861 non-null    int64  
 9   source             861 non-null    object 
dtypes: float64(1), int64(1), object(8)
memory usage: 67.4+ KB


 time_start           0.000000
time_end             0.000000
model                0.000000
launch_place         0.458769
target               0.032520
destroyed_details    0.921022
carrier              0.803717
launched             0.003484
destroyed    

13. (4б) Очистити датафрейм від дублікатів (дублюючихся рядків) та записів, де атака продовжувалась відʼємну або нульову кількість часу (що фізично неможливо). Не забудьте оновити індекс. Скільки рядків було видалено? Як ви це перевірили у коді?

In [122]:
# df.info()
# print()

l1 = len(df)
df = df.drop_duplicates(keep="first")
df = df[df["time_end"] > df["time_start"]].reset_index(drop=False)
l2 = len(df)
print("Rows removed", l1 - l2)


Rows removed 299


14. (3б) Виведіть на екран всі рядки, де є хоч одне пропущене значення.

In [27]:
df[df.isnull().any(axis=1)]

time_start          time_end                          model  \
3          2024-03-23  2024-03-23 08:00                          C-300   
4          2024-03-23        2024-03-23                           ZALA   
13   2024-03-21 03:00  2024-03-21 07:00  Iskander-M/KN-23/X-47 Kinzhal   
14         2024-03-21        2024-03-21                       Supercam   
15         2024-03-21        2024-03-21                         Lancet   
..                ...               ...                            ...   
856  2022-10-01 12:00  2022-10-01 12:00                       Orlan-10   
857        2022-09-30        2022-09-30                       Orlan-10   
858  2022-09-29 23:00  2022-09-29 23:59                 Shahed-136/131   
859        2022-09-28        2022-09-28                           X-59   
860        2022-09-28        2022-09-28             Reconnaissance UAV   

    launch_place                                         target  launched  \
3            NaN                                 Donetsk oblast       4.0   
4            NaN                                          south       2.0   
13           NaN                                    Kyiv oblast       2.0   
14           NaN                                   Odesa oblast       1.0   
15           NaN                                   Odesa oblast       2.0   
..           ...                                            ...       ...   
856          NaN                                        Ukraine       1.0   
857          NaN                                        Ukraine       3.0   
858          NaN               Odesa oblast and Mykolaiv oblast       7.0   
859          NaN  Dnipropetrovsk oblast and Zaporizhzhia oblast       5.0   
860          NaN                                        Ukraine       3.0   

     destroyed                                             source  
3            0  kpszsu/posts/pfbid0Jo2zmM4BzAZfS91x3Eb5Aa6L7AF...  
4            2  PvKPivden/posts/pfbid02rMwctKZrMFpQUXBtK15WCD8...  
13           2  kpszsu/posts/pfbid024ZjSa6LyBpqHEUKgGXtEMfAnpt...  
14           1  PvKPivden/posts/pfbid02ARvTSdz2u6rpdXCqdfjLW6N...  
15           2  PvKPivden/posts/pfbid02ARvTSdz2u6rpdXCqdfjLW6N...  
..         ...                                                ...  
856          1                           watch/?v=375517448001233  
857          3  kpszsu/posts/pfbid0nELeXY4BpHrezv54Lk3gWPZJmcC...  
858          5  kpszsu/posts/pfbid021aK2TqryAuB8gGrDYMTv3R8Xiu...  
859          4  kpszsu/posts/pfbid02sjn5m7kosAfsm5bSUzpPwbLuH3...  
860          3  kpszsu/posts/pfbid0grmBx6A5LbQ7PVTkSx7jTDgg4iU...  

[419 rows x 8 columns]

# V. Data Wrangling

15. (9б) Зʼясуйте, який середній відсоток збиття ракет за весь наведений період.

In [124]:
t_1 = df["launched"].sum()
t_2 = df["destroyed"].sum()
print(t_2 / t_1 * 100)

77.00183374083129


16. (11б) Зʼясуйте 10 найпопулярніших моделей запущених пристроїв за кількістю випущених одиниць за наведений період. Запишіть результат у файл типу json. Скопіюйте зміст файлу та вставте нижче.

In [88]:
t_16 = df["model"].value_counts().nlargest(10)
t_16.to_json('df_json.json', orient = 'table', index = 'true')

17. (10б) Зʼясуйте, у який день було найбільше влучань за весь період (незбитих пристроїв). Виведіть час початку атаки та кількість влучань.

In [158]:
df.info()
print()

t_17 = df[["time_start", "launched", "destroyed"]]
print(t_17)
t_17["hit"] = t_17["launched"] - t_17["destroyed"]

max = t_17["hit"].idxmax()
print(max)
result_row = t_17.iloc[[max]].squeeze()
print("Attack start:", result_row.iloc[0])
print("Max lauches:", result_row.iloc[1])
print("Hit:", result_row.iloc[3])

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 562 entries, 0 to 561
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   index         562 non-null    int64  
 1   time_start    562 non-null    object 
 2   time_end      562 non-null    object 
 3   model         562 non-null    object 
 4   launch_place  415 non-null    object 
 5   target        552 non-null    object 
 6   launched      559 non-null    float64
 7   destroyed     562 non-null    int64  
 8   source        562 non-null    object 
dtypes: float64(1), int64(2), object(6)
memory usage: 39.6+ KB

           time_start  launched  destroyed
0          2024-03-24      29.0         18
1          2024-03-24      28.0         25
2          2024-03-23      34.0         31
3          2024-03-23       4.0          0
4    2024-03-22 02:00       2.0          2
..                ...       ...        ...
557  2022-10-06 09:00       1.0          1
558        2022

C:\Users\Admin\AppData\Local\Temp\ipykernel_11240\2018729190.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  t_17["hit"] = t_17["launched"] - t_17["destroyed"]


18. (15б) Знайдіть найбільш тривалу та найбільш коротку атаки за кожен представлений місяць. Результат отримайте у такому вигляді:

![текст посилання](https://drive.google.com/uc?id=1gISSxR2mXjTEOgoAk_vQnXJeqoL-eA5b)


Запишіть його у файл типу сsv. Після цього виведіть на екран.

In [190]:
t_18 = df
t_18["time_start"] = pd.to_datetime(df['time_start'], format="mixed")
t_18["time_end"] = pd.to_datetime(df['time_end'], format="mixed")
t_18["attack_duration"] = t_18["time_end"] - t_18["time_start"]
t_18['year'] = t_18['time_start'].dt.year
t_18['month'] = t_18['time_start'].dt.month
t_18['duration'] = t_18['time_end'] - t_18['time_start']

longest = t_18.groupby(["year", "month"])["duration"].max().reset_index()
longest = longest.sort_values(["year", "month"], ascending=[False, True]).reset_index(drop=True)

shortest = t_18.groupby(["year", "month"])["duration"].min().reset_index()
shortest = shortest.sort_values(["year", "month"], ascending=[False, True]).reset_index(drop=True)

longest.to_csv('longest.csv', index=False)
shortest.to_csv('shortest.csv', index=False)

print(longest, "\n", shortest)


    year  month        duration
0   2024      1 0 days 12:00:00
1   2024      2 0 days 08:00:00
2   2024      3 0 days 08:00:00
3   2023      1 0 days 08:00:00
4   2023      2 0 days 14:10:00
5   2023      3 0 days 04:30:00
6   2023      4 1 days 10:30:00
7   2023      5 0 days 08:30:00
8   2023      6 0 days 08:00:00
9   2023      7 0 days 08:00:00
10  2023      8 0 days 08:00:00
11  2023      9 0 days 08:00:00
12  2023     10 0 days 09:00:00
13  2023     11 0 days 09:00:00
14  2023     12 0 days 13:00:00
15  2022      9 0 days 00:59:00
16  2022     10 0 days 10:00:00
17  2022     11 0 days 03:00:00
18  2022     12 0 days 09:15:00 
     year  month        duration
0   2024      1 0 days 01:10:00
1   2024      2 0 days 08:00:00
2   2024      3 0 days 04:00:00
3   2023      1 0 days 03:00:00
4   2023      2 0 days 02:05:00
5   2023      3 0 days 02:10:00
6   2023      4 0 days 02:00:00
7   2023      5 0 days 00:30:00
8   2023      6 0 days 02:30:00
9   2023      7 0 days 01:30:00
10  20

# Вітаю! Ви велика(ий) молодець, що впоралась(вся). Похваліть себе та побалуйте чимось приємним. Я Вами пишаюся.